File: zigZagConversion.js
function convert(s, numRows) {
    if (numRows === 1) return s;
    
    const rows = new Array(Math.min(numRows, s.length)).fill("");
    let currentRow = 0;
    let goingDown = false;

    for (const char of s) {
        rows[currentRow] += char;
        if (currentRow === 0 || currentRow === numRows - 1) {
            goingDown = !goingDown;
        }
        currentRow += goingDown ? 1 : -1;
    }

    return rows.join("");
}

File: zigZagConversion.go
package main

import (
	"strings"
)

func convert(s string, numRows int) string {
	if numRows == 1 {
		return s
	}

	rows := make([]strings.Builder, min(numRows, len(s)))
	currentRow := 0
	goingDown := false

	for _, char := range s {
		rows[currentRow].WriteRune(char)
		if currentRow == 0 || currentRow == numRows-1 {
			goingDown = !goingDown
		}
		if goingDown {
			currentRow++
		} else {
			currentRow--
		}
	}

	result := strings.Builder{}
	for _, row := range rows {
		result.WriteString(row.String())
	}
	return result.String()
}
